flink-connectors/flink-connector-pubsub/src/main/java/org/apache/flink/streaming/connectors/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,294 @@
+
+import com.google.auth.Credentials;
+import com.google.cloud.pubsub.v1.AckReplyConsumer;
+import com.google.cloud.pubsub.v1.Subscriber;
+import com.google.pubsub.v1.ProjectSubscriptionName;
+import com.google.pubsub.v1.PubsubMessage;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Optional;
+
+import static com.google.cloud.pubsub.v1.SubscriptionAdminSettings.defaultCredentialsProviderBuilder;
+
+/**
+ * PubSub Source, this Source will consume PubSub messages from a subscription and Acknowledge them on the next checkpoint.
+ * This ensures every message will get acknowledged at least once.
+ */
+public class PubSubSource<OUT> extends MultipleIdsMessageAcknowledgingSourceBase<OUT, String, AckReplyConsumer>
+		implements ResultTypeQueryable<OUT>, ParallelSourceFunction<OUT>, StoppableFunction {
+	private static final Logger LOG = LoggerFactory.getLogger(PubSubSource.class);
+	protected DeserializationSchema<OUT> deserializationSchema;
+	protected SubscriberWrapper subscriberWrapper;
+
+	protected boolean running = true;
+	protected transient volatile SourceContext<OUT> sourceContext = null;
+
+	protected PubSubSource() {
+		super(String.class);
+	}
+
+	protected void setDeserializationSchema(DeserializationSchema<OUT> deserializationSchema) {
+		this.deserializationSchema = deserializationSchema;
+	}
+
+	protected void setSubscriberWrapper(SubscriberWrapper subscriberWrapper) {
+		this.subscriberWrapper = subscriberWrapper;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		subscriberWrapper.initialize();
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesReceivedNotProcessed", subscriberWrapper::amountOfMessagesInBuffer);
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<AckReplyConsumer> ackReplyConsumers) {
+		ackReplyConsumers.forEach(AckReplyConsumer::ack);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		this.sourceContext = sourceContext;
+		subscriberWrapper.start();
+
+		while (subscriberWrapper.isRunning()) {
+			try {
+				Optional.ofNullable(subscriberWrapper.take())
+						.ifPresent(this::processMessage);
+			} catch (InterruptedException e) {
+				LOG.debug("Interrupted - stop or cancel called?");
```

Comment:
My rationale is: Flink will interrupt the thread when it wants the connector to stop/cancel. So when an interrupt is sent, the connector will already be shutting down because of stop()/cancel()  and thus the connector will be breaking out of the while() loop.

Reply:
On shutdown, Flink is calling `cancel()`. If the thread is not shutting down within a certain time, it starts setting the interrupt flag for the thread.
If the thread still ignores the signal (which this thread does), the TaskManager might kill itself by calling System.exit (I think this is disabled by default), because the JVM won't shut down otherwise.
What I want to say is that being interrupted is a last resort, and it means that the thread is not behaving well. If you look at the code of the other connectors, you see that on interrupt, they are just breaking out of the while loop.